Reorganised tutorial directories

eh111eh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Issues #26625

#### Brief description of what is fixed or changed
I created a new directory at `doc/src/tutorials/physics/mechanics`, moved the biomechanics file from `doc/src/tutorials/` to `doc/src/tutorials/physics/`.

@tjstienstra 
@moorepants 

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

eh111eh committed:
changed file location

moorepants commented on doc/src/tutorials/physics/mechanics/DuffingSpring Simulation.py:
```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python
```
You'll need to turn this file into a RestructuredText file with prose and code. You can use the matplotlib plot directive to autogenerate plots. See the biomechanics tutorial pages for examples.

moorepants commented:
I recommend building the docs locally so you can see the results before waiting for the CI to run.

tjstienstra commented on doc/src/conf.py:
```diff
@@ -81,6 +81,8 @@
     "tutorial/matrices": "../tutorials/intro-tutorial/matrices.html",
     "tutorial/manipulation": "../tutorials/intro-tutorial/manipulation.html",
 
+    "tutorials/biomechanics": "tutorials/physics/biomechanics",
```
You should refer to the html pages here, e.g. `"../tutorials/physics/biomechanics/biomechanics.html"`.
Also, make sure to have references in the index files (`index.rst`), e.g. a reference to a tutorials/physics/index.rst file in [`tutorials/index.rst`](https://github.com/sympy/sympy/blob/master/doc/src/tutorials/index.rst).

moorepants replied:
No need to make a redirect for anything that has not yet been released. This stuff is not in SymPy 1.12.

moorepants commented:
@ricdigi Here is one of @eh111eh's PR's. We'd like you and her to help review each other to learn from each other.

eh111eh committed:
Rewrite duffing example in rst

Rewrite duffing example in rst and changed index.rst

moorepants commented:
I just tried building the html docs locally and get:

```
/home/moorepants/src/sympy/doc/src/tutorials/physics/mechanics/duffingspring-model-example.rst:239: WARNING: Exception occurred in plotting duffingspring-model-example-8
 from /home/moorepants/src/sympy/doc/src/tutorials/physics/mechanics/duffingspring-model-example.rst:
Traceback (most recent call last):
  File "/home/moorepants/miniconda/lib/python3.11/site-packages/matplotlib/sphinxext/plot_directive.py", line 552, in _run_code
    exec(code, ns)
  File "<string>", line 10, in <module>
NameError: name 'solve_ivp' is not defined
```

moorepants commented:
I got the example to build with some fixes to the imports and such, but you are not using the DuffingActuator. Isn't the point of the example to show how to use the DuffingAcutator?

moorepants commented:
```diff
diff --git a/doc/src/tutorials/physics/mechanics/duffingspring-model-example.rst b/doc/src/tutorials/physics/mechanics/duffingspring-model-example.rst
index d13b7fdf26..811aa5276a 100644
--- a/doc/src/tutorials/physics/mechanics/duffingspring-model-example.rst
+++ b/doc/src/tutorials/physics/mechanics/duffingspring-model-example.rst
@@ -241,6 +241,11 @@ to check energy conservation.
    :context: reset
 
    >>> import matplotlib.pyplot as plt
+   >>> from scipy.integrate import solve_ivp
+   >>> def duffing_oscillator(t, y, alpha_val, beta_val):
+   ...     x, xdot = y
+   ...     xddot = -beta_val * x - alpha_val * x**3
+   ...     return [xdot, xddot]
    >>> def total_energy(t, y, alpha, beta):
    ...     kinetic = 0.5 * y[1]**2
    ...     potential = 0.5 * beta * y[0]**2 + (alpha / 4) * y[0]**4
@@ -285,4 +290,4 @@ Let's compare with analytical solutions/results from the literature for validati
    >>> plt.axvline(0, color='black', linewidth=0.5)
    >>> plt.grid(True)
    >>> plt.legend(title='Parameter α')
-   >>> plt.show()
\ No newline at end of file
+   >>> plt.show()
```

eh111eh commented:
> I got the example to build with some fixes to the imports and such, but you are not using the DuffingActuator. Isn't the point of the example to show how to use the DuffingAcutator?

You're right, thanks for pointing out. I've rewritten the file to actually use the DuffingActuator. Also, I got the same error you mentioned, but I was unsure of the reason even after some investigation, so I tried to see how it works in CI.

asmeurer commented:
Thanks for adding the redirects. Make sure to test them manually as I don't think there are any checks to make sure they are correct.

moorepants commented:
> Thanks for adding the redirects. Make sure to test them manually as I don't think there are any checks to make sure they are correct.

I don't think we need any redirects because the moved content is not in SymPy 1.12.

moorepants commented:
It may be best to seperate this into two PRs: 1) tutorial directory reorganization to match subpackages and 2) adding a tutorial on acuators (starting with an example using the DuffingActuator).

eh111eh commented:
> It may be best to seperate this into two PRs: 1) tutorial directory reorganization to match subpackages and 2) adding a tutorial on acuators (starting with an example using the DuffingActuator).

I have created a new pull request to add a duffing equation tutorial. Unlike the simulation without the actuator, I encountered some issues which I believe arise from the absolute value used in the force expression. I have detailed this further in PR #26641. Thank you for your help, and I would appreciate any advice on this pull request!

eh111eh committed:
Remove redirect and add physics tutorial

moorepants commented on doc/src/tutorials/physics/index.rst:
```diff
@@ -0,0 +1,28 @@
+.. _physics:
+
+================
+Physics Tutorial
```
```suggestion
Physics Tutorials
```

eh111eh committed:
Update doc/src/tutorials/physics/index.rst

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

moorepants commented:
When I build locally I see:
![image](https://github.com/sympy/sympy/assets/276007/2eb85a33-0427-4e06-a79d-5a359a6fdb8c)
Then if I click the "biomechanics tutorials" header it takes me to the API documentation:
![image](https://github.com/sympy/sympy/assets/276007/17a2069f-6d9c-46ca-b18b-b49e50281674)

So I don't think it is set up correctly. You need:

```
docs/src/tutorials/physics/
--> index.rst
--> biomechanics/
    --> index.rst
    --> tutorial1.rst
    --> tutorial2.rst
```

The table of contents and navigation in those files should lead down that tree and display the table of contents for that tree, not send you to the API docs.

asmeurer commented:
> I don't think we need any redirects because the moved content is not in SymPy 1.12.

It would be good to get this in 1.13 then so that they aren't needed. Is this almost ready?

asmeurer commented:
If not, we can make a quick PR to the release branch that just does the rename.

moorepants commented on doc/src/tutorials/physics/index.rst:
```diff
@@ -0,0 +1,28 @@
+more options and capabilities than what is covered in this tutorial.
+Through insightful examples and exercises, this tutorial demonstrates
+how to incorporate the physics features of SymPy into a multibody dynamics model
+and simulate the resulting equations of motion from the model.
```
The `physics/` subpackage contains many other topics other than multibody dynamics, so this should be reworded to be general enough for all things inside the physics package.

eh111eh replied:
I rewrote as 'Through insightful examples and exercises, this tutorial demonstrates how to incorporate various physics subpackages of SymPy, including but not limited to, multibody dynamics, quantum mechanics, optics, and continuum mechanics, to model and simulate different physical systems and their behaviors.'

asmeurer commented:
Oh apparently this PR is just renames. So we should get it fixed so it can be backported.

moorepants commented:
Yes, this needs to go into 1.13 ASAP. We discussed this with @eh111eh in our GSoC meeting yesterday.

eh111eh committed:
Update physics, biomechanics set up

eh111eh commented:
> When I build locally I see: ![image](https://private-user-images.githubusercontent.com/276007/337148587-2eb85a33-0427-4e06-a79d-5a359a6fdb8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2NTk1ODAsIm5iZiI6MTcxNzY1OTI4MCwicGF0aCI6Ii8yNzYwMDcvMzM3MTQ4NTg3LTJlYjg1YTMzLTA0MjctNGUwNi1hNzlkLTVhMzU5YTZmZGI4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNlQwNzM0NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNTM5NDE2MzA2ZjMzNzY3MGYzMzA4ZTNlNzAwYmEwMzhkOGY3ZjE1NGFkYzZhMDcxYjhkOWQ0MjcxYjFhYjJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.m7x7DFAFCViYxfc_R9hgcJ9ztaKXJFC9mSX5qOaOl68) Then if I click the "biomechanics tutorials" header it takes me to the API documentation: ![image](https://private-user-images.githubusercontent.com/276007/337148810-17a2069f-6d9c-46ca-b18b-b49e50281674.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2NTk1ODAsIm5iZiI6MTcxNzY1OTI4MCwicGF0aCI6Ii8yNzYwMDcvMzM3MTQ4ODEwLTE3YTIwNjlmLTZkOWMtNDZjYS1iMThiLWI0OWU1MDI4MTY3NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNlQwNzM0NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDE0MDU1YTA2Nzk1NDZkYjFhMGEyOTVjYzI5ODZiYTkyMjM4ZjVlMjU1NmYwZmVjNzE5ODE2MjY4M2VlMWRkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.e6gZIxmCufVlwU2IlZkdN82Ja71dCOaM88cr6LVvzCQ)
> 
> So I don't think it is set up correctly. You need:
> 
> ```
> docs/src/tutorials/physics/
> --> index.rst
> --> biomechanics/
>     --> index.rst
>     --> tutorial1.rst
>     --> tutorial2.rst
> ```
> 
> The table of contents and navigation in those files should lead down that tree and display the table of contents for that tree, not send you to the API docs.

For now,
- physics tutorials
<img width="826" alt="Screenshot 2024-06-06 at 15 40 51" src="https://github.com/sympy/sympy/assets/97640870/6ff7cecb-c892-48fc-a390-bc50aae1eed9">

- biomechanics
<img width="826" alt="Screenshot 2024-06-06 at 15 40 57" src="https://github.com/sympy/sympy/assets/97640870/a10049b7-0d75-4ef6-9005-c89ee0d33c1b">

And the directory is:
```
docs/src/tutorials/physics/
--> index.rst
--> biomechanics/
   --> index.rst
   --> biomechanics.rst
   --> biomechanical-model-example.rst
```

So I hope it now works as expected

tjstienstra commented on doc/src/tutorials/physics/index.rst:
```diff
@@ -0,0 +1,24 @@
+Physics Tutorials
+=================
+
+This tutorial aims to introduce SymPy to users who have not previously
+used its physics features. The functionalities presented here have many
+more options and capabilities than what is covered in this tutorial.
+Through insightful examples and exercises, this tutorial demonstrates
+how to incorporate various physics subpackages of SymPy, including but not limited to,
+multibody dynamics, quantum mechanics, optics, and continuum mechanics, to model and simulate
+different physical systems and their behaviors.
+
+Feedback on this tutorial, or on SymPy in general is always welcome. Just
+write to our `mailing list
+<https://groups.google.com/forum/?fromgroups#!forum/sympy>`_.
```
Can you also change `tutorial` to plural in these paragraphs?

tjstienstra commented:
You'll have to update the [blacklist in bin/doctest](https://github.com/sympy/sympy/blob/da2d56142230563bcfe50c564980e7d07fcc0cc9/bin/doctest#L14) to ensure that `tests/doctests-latest` doesn't fail.

moorepants commented:
I built locally and it is all correct now. You'll have to check out the failing CI jobs.

moorepants commented on doc/src/tutorials/physics/biomechanics/index.rst:
```diff
@@ -0,0 +1,13 @@
+   :maxdepth: 2
+
+   biomechanical-model-example.rst
+   biomechanics.rst
```
It would preferable to have the intro before the example.

eh111eh replied:
I wrote as follows, and if there's anything else to mention, please let me know!
```
These tutorials provide a comprehensive guide on using SymPy for
biomechanical simulations and analysis. We cover various models, including
a human arm moving a lever, forces produced by muscles, and
tendons using Hill-type muscle models.

In these tutorials, you can expect:

- Model description: Detailed explanation of each biomechanical model.

- Variables and Kinematics: Definition of necessary variables and kinematics equations essential for modeling.

- Modeling: Step-by-step process of constructing the biomechanical models.

- Equations of Motion: Derivation and analysis of the equations of motion for the systems.
```

moorepants replied:
I meant swap lines 25 and 26 so that the Intro tutorial is list first in the table of contents. The reader should look at the intro tutorial before the example (or that should be the recommendation).

asmeurer commented:
For the doctests you need to update these https://github.com/sympy/sympy/blob/da2d56142230563bcfe50c564980e7d07fcc0cc9/sympy/testing/runtests.py#L694

eh111eh committed:
Update index, doctest, and runtests

- Match the length === and the title
- Add intro to biomechanics tutorials
- Plural
- Change runtest.py and bin/doctest to pass doctest
- Change the duplicate label

eh111eh commented on doc/src/tutorials/index.rst:
```diff
@@ -17,14 +17,12 @@ If you are new to SymPy, start here.
-Biomechanics Tutorials
-======================
+:ref:`Physics Tutorial <physics>`
+==========================================
```
Sure, I just changed the file so now the lengths of === and the title are the same
```
:ref:`Physics Tutorial <physics_tutorials>`
===========================================
```

eh111eh commented on doc/src/tutorials/physics/biomechanics/index.rst:
```diff
@@ -0,0 +1,13 @@
+Biomechanics Tutorials
+======================
+
+This tutorial focuses on biomechanics.
```
I rewrote in plural and add some intro
```
These tutorials provide a comprehensive guide on using SymPy for
biomechanical simulations and analysis. We cover various models, including
a human arm moving a lever, forces produced by muscles, and
tendons using Hill-type muscle models.

In these tutorials, you can expect:

- Model description: Detailed explanation of each biomechanical model.

- Variables and Kinematics: Definition of necessary variables and kinematics equations essential for modeling.

- Modeling: Step-by-step process of constructing the biomechanical models.

- Equations of Motion: Derivation and analysis of the equations of motion for the systems.
```

moorepants commented:
Also fix the authors file so that CI action passes.

eh111eh committed:
Update .mailmap

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into example-location

eh111eh committed:
Update .mailmap

eh111eh committed:
Update index.rst

moorepants commented:
@oscarbenjamin we'd like this in the SymPy 1.13 release so we don't have to use redirects to the moved docs. Is that possible? If so, what should we do?

oscarbenjamin commented:
I've just pushed the tag for 1.13.0rc1 but I think it is okay to make this change between rc1 and the final release.

You would need to open a PR against the 1.13 release branch backporting the changes here. You can use `git cherry-pick` to apply the commits from here to a new PR branch based off of 1.13. Don't cherry-pick the merge commit though.

moorepants commented:
I'm struggling to figure out how to cherry pick this set of commits. I have tried picking the commits spanning the first to last and individually selecting all commits in this PR as well as leaving out the merge of master into this PR. I get into loops of merge conflicts and failure to apply the cherry pick. I haven't found a successful combination for a succeeding cherrypick. Any tips?

moorepants commented:
Maybe this is sufficient: https://github.com/sympy/sympy/pull/26680

oscarbenjamin commented:
It would have been better to squash the commits here in the first place (see gh-18548). There is no reason that we need 10 commits for this. It is much easier to cherry-pick a single commit with no intermediate merges.
